core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1885,6 +1885,14 @@ private[spark] class DAGScheduler(
               mapOutputTracker.
                 unregisterMergeResult(shuffleId, reduceId, bmAddress, Option(mapIndex))
             }
+          } else {
+            // Unregister the merge result of <shuffleId, reduceId> if
+            // the FetchFailed event contains a mapIndex of -1, and is not a
+            // MetaDataFetchException which is signified by bmAddress being null
+            if (bmAddress != null && pushBasedShuffleEnabled) {
```

Comment:
Why do we need `pushBasedShuffleEnabled` check here ?

Reply:
we dont need that check here, removed

Reply:
I think we should check for `pushBasedShuffleEnabled` here just for consistency with lines 1882-1886. There aren't going to be any merge results if push-based shuffle is not enabled but since we do it at the other places we should do it here as well.

Reply:
I am trying to understand, can we have a case where `mapIndex == -1` and `bmAddress != null` for non-push based shuffle ? If no, we should add an assertion there (not drop the check entirely).

Note, L1882 has the check because the earlier check is for `mapIndex != -1` - which can be satisfied for all shuffles

Reply:
added the assertion